chapters/operatorsandexpressions.tex

```diff
@@ -514,13 +511,16 @@ \subsection{Numeric Functions and Conversion Functions}\label{numeric-functions-
   \lstinline!Boolean leftJustified = true!: If true, the converted result is left justified in the string; if false it is right justified in the string.
 \item
   \lstinline!Integer significantDigits = 6!: Number of significant digits in the result string.
+  Only allowed when formatting a \lstinline!Real! value.
```

Comment:
Should we clarify that String(4, format="4d") is allowed here as well? (Since the integer is converted to real.)

Reply:
To me it wasn't even clear that implicit conversion to `Real` would happen in this case, due to the existence of a form of `String` with matching type in the position of the `4`.  It would be good to get language group agreement on this, and clarify one way or another.

Reply:
By the way, I had also expected this to be an error: `String(4, significantDigits = 4)`

Reply:
> By the way, I had also expected this to be an error: `String(4, significantDigits = 4)`

I would expect the standard type coercion ([10.6.13](https://specification.modelica.org/master/arrays.html#standard-type-coercion)) to be applied in that case, so that it's converted to `String(4.0, significantDigits = 4)`.

Reply:
While I can see that 10.6.13 could be interpreted this way, I don't think it's obvious that it must be interpreted this way due to the fact that `String` has another form that immediately accepts the `Integer` value.  If we'd say this is allowed, it feels contradictory (at least on the surface) to state that `significantDigits` is only allowed for `Real`.

If there's room for interpretation, why not go with the simpler and more strict view that `significantDigits` isn't allowed since the first argument is an `Integer`?  Do we fear that anyone would actually miss the possibility to pass an `Integer` expression in combination with `significantDigits`?

Reply:
The `Integer` overload of `String` is not a valid match since it doesn't have the  `significantDigits` parameter, so there's only one valid overload in that case which is the one that takes a `Real`.

Reply:
Sure, that's the other interpretation.  I'm just not sure it has to be the only possible interpretation at the moment.  I mean, do we have any other examples to look at where the presence of a named argument determines how to perform type conversion on other arguments in a function call?

Reply:
> To me it wasn't even clear that implicit conversion to `Real` would happen in this case, due to the existence of a form of `String` with matching type in the position of the `4`. It would be good to get language group agreement on this, and clarify one way or another.

To me that was given due to "d" and "x" being discussed. Saying that `String(4.0, format="5d")` is allowed, but not `String(4, format="5d")` would be weird.

Reply:
Aha, I think I can see the source of the confusion now:
> For integral values it is also allowed to use the ‘d’, ‘i’, ‘o’, ‘x’, ‘X’, ‘u’, and ‘c’ type field specifiers

That is, we actually seem to be missing the `Integer` case in the synopsis:
```
String(i, format = s)
```

Should we add it, and see how that affects the rest of the presentation?

Reply:
> > I mean, do we have any other examples to look at where the presence of a named argument determines how to perform type conversion on other arguments in a function call?
> 
> Yes, [overloaded operators](https://specification.modelica.org/master/overloaded-operators.html#matching-function).

What I see in https://specification.modelica.org/master/overloaded-operators.html#overloaded-string-conversions is that one should start by looking at the type of the first argument, and then find a matching function based on the remaining arguments.

Reply:
I would interpret it more that if the first argument is one of the builtin types then the matching `String` overload is selected. How that is done is not actually specified, but the only rules we really have for that is the ones in 14.2.

However, I realized that there's one issue with the rules in 14.2 which is that it says that `Integer` and `Real` should be treated as the same type, which is not actually what we implement in OpenModelica. We prefer an exact match over a match that does type casting, otherwise it doesn't work for functions that have overloads for both `Integer` and `Real`. But that's mostly because we use this for any builtin function that's overloaded, which is not strictly specified.

So perhaps this is beyond the scope of this ticket and we need to resolve how overloaded builtin functions should actually behave first.

Reply:
Also, adding `String(i, format = s)` would be necessary if we want different sets of conversion specifiers for `Integer` and `Real`. But even if we do that I think we still need to clarify how overloading of builtin functions should work.

Reply:
> but for compatibility allowing both int and double formatting for both integer and real arguments, so that String(40000000, format="g") and String(4.0, format="d") both work.

I don't see a clear need to add support for `String(4.0, format="d")`, as this only corresponds to this strange parenthesized remark:
> (for `Real` values a tool may round, truncate or use a different format if the integer conversion specifiers are used)

I mean, it hasn't even been clearly defined what to expect of this mismatched situation in the past, so wouldn't it be better to instead go in the other direction and just drop this vague remark?  As the consequence of dropping it would be that it isn't allowed at all, backwards compatibility fighters might prefer to add back a deprecation statement:
> A deprecated feature is that a `Real` value can also be formatted using one of the format specifiers for `Integer` values, although exactly what to output in this case is not defined.

Reply:
A nice way to interpret `String(42, format="g")` would be that the `Integer` case of `String` is responsible for processing the `"g"` format by conversion to `double` _inside_ the operator's implementation. Thus, there is no need for an implicit conversion to `Real` already in the argument _given to_ the operator.

The same approach could be taken if we wanted to support `String(42, significantDigits = 4)`: Just add extends the `Integer` value case to also accept the `significantDigits` option.

With these two steps, maybe we wouldn't need to sort out the general case of overloaded operator resolution when named arguments are present or when some argument's value could determine the expected type of another argument.  However, I am still not sure it's a good idea to allow `String(42, significantDigits = 4)`, but getting rid of this would either require diving into the operator overload question, or require a special statement about how the overloading of `String` works.